tables/endpointsecurity/src/fileeventstableplugin.cpp

```diff
@@ -0,0 +1,171 @@
+#include "fileeventstableplugin.h"
+
+#include <chrono>
+#include <mutex>
+
+
+namespace zeek {
+struct FileEventsTablePlugin::PrivateData final {
+  PrivateData(IZeekConfiguration &configuration_, IZeekLogger &logger_)
+      : configuration(configuration_), logger(logger_) {}
+
+  IZeekConfiguration &configuration;
+  IZeekLogger &logger;
+
+  RowList row_list;
+  std::mutex row_list_mutex;
+  std::size_t max_queued_row_count{0U};
+};
+
+Status FileEventsTablePlugin::create(Ref &obj,
+                                     IZeekConfiguration &configuration,
+                                     IZeekLogger &logger) {
+  obj.reset();
+
+  try {
+    auto ptr = new FileEventsTablePlugin(configuration, logger);
+    obj.reset(ptr);
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
+
+  } catch (const Status &status) {
+    return status;
+  }
+}
+
+FileEventsTablePlugin::~FileEventsTablePlugin() {}
+
+const std::string &FileEventsTablePlugin::name() const {
+  static const std::string kTableName{"file_events"};
+
+  return kTableName;
+}
+
+const FileEventsTablePlugin::Schema &FileEventsTablePlugin::schema() const {
+  // clang-format off
+  static const Schema kTableSchema = {
+    { "timestamp", IVirtualTable::ColumnType::Integer },
+    { "type", IVirtualTable::ColumnType::String },
+    { "parent_process_id", IVirtualTable::ColumnType::Integer },
+    { "orig_parent_process_id", IVirtualTable::ColumnType::Integer },
+    { "process_id", IVirtualTable::ColumnType::Integer },
+    { "user_id", IVirtualTable::ColumnType::Integer },
+    { "group_id", IVirtualTable::ColumnType::Integer },
+    { "platform_binary", IVirtualTable::ColumnType::Integer },
+    { "signing_id", IVirtualTable::ColumnType::String },
+    { "team_id", IVirtualTable::ColumnType::String },
+    { "cdhash", IVirtualTable::ColumnType::String },
+    { "path", IVirtualTable::ColumnType::String },
+    { "file_path", IVirtualTable::ColumnType::String }
+  };
+  // clang-format on
+
+  return kTableSchema;
+}
+
+Status FileEventsTablePlugin::generateRowList(RowList &row_list) {
+  std::lock_guard<std::mutex> lock(d->row_list_mutex);
+
+  row_list = std::move(d->row_list);
+  d->row_list = {};
+
+  return Status::success();
+}
+
+Status FileEventsTablePlugin::processEvents(
+    const IEndpointSecurityConsumer::EventList &event_list) {
+  RowList generated_row_list;
```

Comment:
Could you add a comment explaining why we first buffer all rows into a local list and only then publish them in the final `row_list`? This e.g., causes us potentially require us unbounded extra storage for this function, and then later us holding a lock on the table for a long time.

Naively it would seem to make more sense to generate a row, acquire a lock for the final table, publish there; dropping events beyond capacity would still happen after that loop but would only require a lock if we actually are beyond capacity (e.g., check first, acquire lock if beyond capacity, then compute number of actual rows to remove under lock).

Reply:
Yes, you are correct about the table update workflow. I will implement row addition in the way you described in my next commit.

Reply:
I have incorporated the row insertion workflow as suggested by you in the above comment.